homeassistant/components/hassio/__init__.py

```diff
@@ -898,47 +908,75 @@ async def force_info_update_supervisor(self) -> None:
-            self.hassio.get_core_info(),
-            self.hassio.get_core_stats(),
-            self.hassio.get_supervisor_info(),
-            self.hassio.get_supervisor_stats(),
-            self.hassio.get_os_info(),
+            hassio.get_info(),
+            hassio.get_core_info(),
+            hassio.get_core_stats(),
+            hassio.get_supervisor_info(),
+            hassio.get_supervisor_stats(),
+            hassio.get_os_info(),
         )
 
-        all_addons = self.hass.data[DATA_SUPERVISOR_INFO].get("addons", [])
-        started_addons = [
-            addon for addon in all_addons if addon[ATTR_STATE] == ATTR_STARTED
-        ]
-        stats_data = await asyncio.gather(
-            *[self._update_addon_stats(addon[ATTR_SLUG]) for addon in started_addons]
-        )
-        self.hass.data[DATA_ADDONS_STATS] = dict(stats_data)
-        self.hass.data[DATA_ADDONS_CHANGELOGS] = dict(
-            await asyncio.gather(
-                *[
-                    self._update_addon_changelog(addon[ATTR_SLUG])
-                    for addon in all_addons
-                ]
-            )
-        )
-        self.hass.data[DATA_ADDONS_INFO] = dict(
-            await asyncio.gather(
-                *[self._update_addon_info(addon[ATTR_SLUG]) for addon in all_addons]
+        _addon_data = data[DATA_SUPERVISOR_INFO].get("addons", [])
+        all_addons: list[str] = []
+        started_addons: list[str] = []
+        for addon in _addon_data:
+            slug = addon[ATTR_SLUG]
+            all_addons.append(slug)
+            if addon[ATTR_STATE] == ATTR_STARTED:
+                started_addons.append(slug)
+        #
+        # Update add-on info if its the first update or
+        # there is at least one entity that needs the data.
+        #
+        # When entities are added they call async_enable_addon_updates
+        # to enable updates for the endpoints they need via
+        # async_added_to_hass. This ensures that we only update
+        # the data for the endpoints that are needed to avoid unnecessary
+        # API calls since otherwise we would fetch stats for all add-ons
+        # and throw them away.
+        #
+        enabled_updates_by_addon = self._enabled_updates_by_addon
+        for data_key, update_func, enabled_key, wanted_addons in (
+            (
+                DATA_ADDONS_STATS,
+                self._update_addon_stats,
+                ADDON_UPDATE_STATS,
+                started_addons,
+            ),
+            (
+                DATA_ADDONS_CHANGELOGS,
+                self._update_addon_changelog,
+                ADDON_UPDATE_CHANGELOG,
+                all_addons,
+            ),
+            (DATA_ADDONS_INFO, self._update_addon_info, ADDON_UPDATE_INFO, all_addons),
+        ):
+            data.setdefault(data_key, {}).update(
+                dict(
+                    await asyncio.gather(
+                        *[
+                            update_func(slug)
```

Comment:
This all seems very complicated. Wouldn't we just want to instantiate an update coordinator per add-on and have it use its own flow for updating ?

Reply:
I did that first and it got much more complicated because the problem is that each entity can use multiple endpoints to gather the data it needs. 

For example the update entity needs the changelog and the info endpoints so we end up with multiple coordinators feeding data into the entity